Python/Product/Analysis/ProjectEntry.cs

```diff
@@ -342,15 +316,101 @@ where fileName.StartsWith(pathPrefix)
             }
         }
 
-        #endregion
-
         internal void Enqueue() {
             _unit.Enqueue();
         }
 
         public void AggregatedInto(AggregateProjectEntry into) {
             _aggregates.Add(into);
         }
+
+        public IEnumerable<int> DocumentParts => _buffers.Keys.AsLockedEnumerable(_buffers);
+
+        public Stream ReadDocumentBytes(int part, out int version) {
+            lock (_buffers) {
+                if (_buffers[part].Version >= 0) {
+                    version = _buffers[part].Version;
+                    return EncodeToStream(_buffers[part].Text, Encoding.UTF8);
+                } else if (part == 0) {
+                    var s = PathUtils.OpenWithRetry(FilePath, FileMode.Open, FileAccess.Read, FileShare.Read);
+                    version = s == null ? -1 : 0;
+                    return s;
+                }
+                version = -1;
+                return null;
+            }
+        }
+
+        internal static Stream EncodeToStream(StringBuilder text, Encoding encoding, int chunkSize = 4096) {
```

Comment:
I think it can be private

Reply:
It needs tests (guess that's my job this morning)